test_utils/utils/create_partition.py

```diff
@@ -0,0 +1,85 @@
+try:
+	import frappe
+	from frappe.utils import get_table_name
+except Exception as e:
+	raise (e)
+
+
+def primary_key_exists(table_name):
+	try:
+		result = frappe.db.sql(
+			f"""
+		SELECT COUNT(*)
+		FROM information_schema.TABLE_CONSTRAINTS
+		WHERE TABLE_NAME = '{table_name}'
+		AND CONSTRAINT_TYPE = 'PRIMARY KEY';
+		"""
+		)
+		return result[0][0] > 0
+	except Exception as e:
+		print(f"Error checking primary key existence: {e}")
+		return False
+
+
+def modify_primary_key(table_name, date_field):
+	try:
+		if primary_key_exists(table_name):
+			drop_pk_sql = f"""
+			ALTER TABLE `{table_name}`
+			DROP PRIMARY KEY;
+			"""
+			frappe.db.sql(drop_pk_sql)
+
+			pk_columns = f"name, {date_field}"
+			add_pk_sql = f"""
+			ALTER TABLE `{table_name}`
+			ADD PRIMARY KEY ({pk_columns});
+			"""
+			frappe.db.sql(add_pk_sql)
+			frappe.db.commit()
+			print(f"Primary key modified in table {table_name} to include columns: {pk_columns}")
+	except Exception as e:
+		print(f"Error modifying primary key: {e}")
+
+
+def create_partition():
+	"""
+	partition_doctypes = {
+	        "Sales Order": {
+	                "date_field": "transaction_date",
+	        },
+	        "Sales Invoice": {
+	                "date_field": "posting_date",
+	        },
+	"""
+	partition_doctypes = frappe.get_hooks("partition_doctypes")
+	fiscal_years = frappe.get_all(
+		"Fiscal Year",
+		fields=["name", "year_start_date", "year_end_date"],
+		order_by="year_start_date ASC",
+	)
+
+	for doctype, settings in partition_doctypes.items():
+		table_name = get_table_name(doctype)
```

Comment:
I propose we also get child tables and add the same field if it doesn't exist (it shouldn't) and constraint to them. In the `doc_event` we can synchronize the date into these tables as well. It doesn't need to be a docfield, I think, but it might need to be. If that's the case, then we can create a custom hidden field with a Frappe API.